### PR TITLE
[Self-service error codes] Do not return error code id and definite_answer in metadata for security sensitive errors

### PR DESCRIPTION
**Pull request by tudor-da**
This PR fixes the current situation where the error code id and definite_answer are returned in the gRPC metadata even for security sensitive error categories. A specific test is written for this in `ErrorCodeSpec`

Additionally, this PR enforces the test assertions (by asserting the metadata in `ErrorFactoriesSpec`) and removes some redundant information for the error contexts

- Hint: see fields renames with `_` prefix 

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by fabiotudone-da — APPROVED** [review]

**Review by pbatko-da — APPROVED** [review]

File: ledger/error/src/main/scala/com/daml/error/definitions/IndexErrors.scala [hunk 1]
```diff
@@ -56,21 +53,34 @@ object IndexErrors extends IndexErrorGroup {
...
+      override val throwableO: Option[Throwable] = None,
+  )(implicit
+      code: ErrorCode,
+      loggingContext: ContextualizedErrorLogger,
+  ) extends LoggingTransactionErrorImpl(cause, throwableO) {
+    override def asGrpcErrorFromContext(implicit
+        contextualizedErrorLogger: ContextualizedErrorLogger
+    ): IndexDbException = {
+      val err = super.asGrpcErrorFromContext(contextualizedErrorLogger)
+      IndexDbException(err.getStatus, err.getTrailers)
```
Comment (pbatko-da) [inline ledger/error/src/main/scala/com/daml/error/definitions/IndexErrors.scala:80]: Out of interest, is `IndexDbException` name leaked to the user of the API? (I.e. can users tell a difference between getting `StatusRuntimeException` and `IndexDbException`?)

Comment (tudor-da) [reply ledger/error/src/main/scala/com/daml/error/definitions/IndexErrors.scala:80]: I don't think so. What gets over the wire is the status and metadata. JVM type information should be ignored I believe

Comment (fabiotudone-da) [reply ledger/error/src/main/scala/com/daml/error/definitions/IndexErrors.scala:80]: Possibly in logs but that shouldn't be an issue I think

File: ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala [hunk 2]
```diff
@@ -808,19 +819,17 @@ object LedgerApiErrors extends LedgerApiErrorGroup {
...
-          message: String,
+          override val cause: String,
           ledger_time: Instant,
           ledger_time_lower_bound: Instant,
           ledger_time_upper_bound: Instant,
       )(implicit loggingContext: ContextualizedErrorLogger)
-          extends LoggingTransactionErrorImpl(
-            cause = s"Invalid ledger time: $message"
-          )
+          extends LoggingTransactionErrorImpl(cause = cause)
```
Comment (mziolekda) [inline ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:827]: Why do you need this assignment? Doesn't `LoggingTransactionErrorImpl(cause)` suffice?

**Review by mziolekda — APPROVED** [review]

File: ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/error/src/main/scala/com/daml/error/definitions/LedgerApiErrors.scala:827]: usually I add named arguments when the args list has default values for some arguments.

**Pull request merged**